File: day02/CubeConundrum.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.Scanner;

public class CubeConundrum {
    private static ArrayList<String> lines = new ArrayList<>();
    private static final int RED = 12;
    private static final int GREEN = 13;
    private static final int BLUE = 14;

    public static void main(String[] args) {
        String filePath = "input.txt";
        File file = new File(filePath);

        readInput(file, filePath);
        calculateCalibration();
    }

    private static void readInput(File file, String filePath) {
        try {
            Scanner scanner = new Scanner(file);
            
            while (scanner.hasNextLine()) {
                lines.add(scanner.nextLine());
            }

            scanner.close();
        } catch (FileNotFoundException e) {
            System.out.println("File not found: " + filePath);
        }
    }

    private static void calculateCalibration() {
        int answer = lines.stream().map(string -> {
            String[] split = string.split(": ");
            String[] sets = split[1].split("; ");
            Boolean valid = true;

            for (String set : sets) {
                String[] colors = set.split(", ");

                for (String color : colors) {
                    valid &= validColorSet(color);
                }
            }

            return valid ? getId(split[0]) : 0;
        }).reduce(0, (first, second) -> first + second);

        System.out.println(answer);
    }

    private static Boolean validColorSet(String color) {
        Boolean valid = true;
        
        String[] colorSplit = color.split(" ");
        int colorValue = Integer.parseInt(colorSplit[0]);

        if (colorSplit[1].equals("red")) {
            valid &= colorValue <= RED;
        } else if (colorSplit[1].equals("green")) {
            valid &= colorValue <= GREEN;
        } else if (colorSplit[1].equals("blue")) {
            valid &= colorValue <= BLUE;
        }

        return valid;
    }

    private static int getId(String gameName) {
        return Integer.parseInt(gameName.split(" ")[1]);
    }

}

File: 2023/day02/cube_conundrum.py
import re

RED = 12
GREEN = 13
BLUE = 14

with open("input.txt") as f:
    lines = f.readlines()
    total = 0
    for line in lines:
        game_name, sets = line.split(':')
        set = sets.split(';')
        valid = True
        for colors in set:
            for color in colors.split(','):
                cube = color.strip()
                if cube:
                    if 'red' in cube:
                        valid = valid and (int(re.findall(r'\d+', cube)[0]) <= RED)
                    elif 'green' in cube:
                        valid = valid and (int(re.findall(r'\d+', cube)[0]) <= GREEN)
                    elif 'blue' in cube:
                        valid = valid and (int(re.findall(r'\d+', cube)[0]) <= BLUE)
        if valid:
            total += int(re.findall(r'\d+', game_name)[0])
    print(total)
